File: src/main/java/cloud/robinzon/backend/db/vm/resources/history/VmHistory.java
package cloud.robinzon.backend.db.vm.resources.history;

import cloud.robinzon.backend.db.fm.resources.FmEntity;
import cloud.robinzon.backend.db.vm.resources.VmEntity;
import cloud.robinzon.backend.security.user.resources.UserEntity;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import java.sql.Timestamp;

@SuppressWarnings("unused")
@Entity
@Getter
@NoArgsConstructor
@IdClass(VmHistoryKey.class)
public class VmHistory {

    @Id
    @ManyToOne
    @JoinColumn
    private VmEntity vmEntity;

    @Id
    @CreationTimestamp
    private Timestamp timestamp;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(nullable = false)
    private int cpu;

    @Column(nullable = false)
    private int ram;

    @Column(nullable = false)
    private int ssd;

    @Column(nullable = false)
    private int hdd;

    @Column(nullable = false)
    private boolean running;

    @Column(nullable = false, length = 50)
    private String title;

    @Column(nullable = false)
    private String description;

    @ManyToOne
    @JoinColumn(nullable = false)
    private FmEntity fmEntity;

    @ManyToOne
    @JoinColumn(nullable = false)
    private UserEntity changeBy;

    @Column(nullable = false)
    private boolean deleted;

    public VmHistory(VmEntity vmEntity,
                     String name,
                     Integer cpu,
                     Integer ram,
                     Integer ssd,
                     Integer hdd,
                     Boolean running,
                     FmEntity fmEntity,
                     UserEntity changeBy,
                     boolean isNew) {
        this.vmEntity = vmEntity;
        this.name = name;
        this.cpu = cpu;
        this.ram = ram;
        this.ssd = ssd;
        this.hdd = hdd;
        this.running = running;
        this.title = isNew ? "" : vmEntity.getTitle();
        this.description = isNew ? "" : vmEntity.getDescription();
        this.fmEntity = fmEntity;
        this.changeBy = changeBy;
        this.deleted = false;
    }

    public VmHistory(VmEntity vmEntity,
                     String title,
                     String description,
                     UserEntity changeBy) {
        this.vmEntity = vmEntity;
        this.name = vmEntity.getName();
        this.cpu = vmEntity.getCpu();
        this.ram = vmEntity.getRam();
        this.ssd = vmEntity.getSsd();
        this.hdd = vmEntity.getHdd();
        this.running = vmEntity.isRunning();
        this.title = title;
        this.description = description;
        this.fmEntity = vmEntity.getFmEntity();
        this.changeBy = changeBy;
        this.deleted = false;
    }

    public VmHistory(VmEntity vmEntity,
                     UserEntity changeBy) {
        this.vmEntity = vmEntity;
        this.name = vmEntity.getName();
        this.cpu = vmEntity.getCpu();
        this.ram = vmEntity.getRam();
        this.ssd = vmEntity.getSsd();
        this.hdd = vmEntity.getHdd();
        this.running = vmEntity.isRunning();
        this.title = vmEntity.getTitle();
        this.description = vmEntity.getDescription();
        this.fmEntity = vmEntity.getFmEntity();
        this.changeBy = changeBy;
        this.deleted = true;
    }

}

File: src/main/java/cloud/robinzon/backend/data/vm/resources/history/VmHistory.kt
package cloud.robinzon.backend.data.vm.resources.history

import cloud.robinzon.backend.data.fm.resources.FmEntity
import cloud.robinzon.backend.data.vm.resources.VmEntity
import cloud.robinzon.backend.security.user.resources.UserEntity
import jakarta.persistence.*
import java.sql.Timestamp

@Entity
class VmHistory(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @ManyToOne
    @JoinColumn(nullable = false)
    val entity: VmEntity? = null,

    @ManyToOne
    @JoinColumn(nullable = false)
    val changeBy: UserEntity? = null,

    @Column(nullable = false)
    val timestamp: Timestamp = Timestamp(System.currentTimeMillis()),

    @Column(nullable = false, columnDefinition = "boolean default false")
    val deleted: Boolean = false,

    @Column(nullable = false, length = 50)
    val title: String = "",

    @Column(nullable = false)
    val description: String = "",

    @Column(nullable = false, length = 100)
    val name: String = "",

    @Suppress("unused") @Column(nullable = false)
    val cpu: Int = 0,

    @Suppress("unused") @Column(nullable = false)
    val ram: Int = 0,

    @Suppress("unused") @Column(nullable = false)
    val ssd: Int = 0,

    @Suppress("unused") @Column(nullable = false)
    val hdd: Int = 0,

    @Suppress("unused") @Column(nullable = false)
    val running: Boolean = false,

    @Suppress("unused") @ManyToOne
    @JoinColumn(nullable = false)
    val fmEntity: FmEntity? = null

) {
    constructor(entity: VmEntity, changeBy: UserEntity) : this(
        null,
        entity,
        changeBy,
        Timestamp(System.currentTimeMillis()),
        entity.deleted,
        entity.title,
        entity.description,
        entity.name,
        entity.cpu,
        entity.ram,
        entity.ssd,
        entity.hdd,
        entity.running,
        entity.fmEntity
    )
}